adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -463,17 +474,103 @@ private void handleConnectionRequestResult(final MqttEndpoint endpoint,
+                        createCommandConsumer(endpoint, cmdSub).setHandler(h -> {
+                            if (h.succeeded()) {
+                                h.result().flow(1);
+                                sendTtdEvent(cmdSub.getTenant(), cmdSub.getDeviceId(), authenticatedDevice, -1);
+                            } else {
+                                LOG.error("command consumer could not be created for [tenant:{}, device:{}]: {}",
+                                        cmdSub.getTenant(), cmdSub.getDeviceId(), h.cause());
+                            }
+                        });
+                        // if called from unauthenticated device subscription overwrite the close
+                        // handler with now known tenant and device
+                        if(authenticatedDevice==null) {
+                            endpoint.closeHandler(v -> {
+                                close(endpoint, new Device(cmdSub.getTenant(), cmdSub.getDeviceId()));
+                            });
+                        }
+                    } else {
+                        LOG.error("denied subscription for exactly once");
+                    }
+                }
+            } else {
+                LOG.error("subscription for multiple topics is not allowed");
+            }
+            if (grantedQosLevels.isEmpty()) {
+                grantedQosLevels.add(MqttQoS.FAILURE);
+                endpoint.subscribeAcknowledge(subscribe.messageId(), grantedQosLevels);
+            }
+        }).unsubscribeHandler(unsubscribe -> {
+            if (unsubscribe.topics().size() == 1) {
+                final String topic = unsubscribe.topics().get(0);
+                final CommandSubscription cmdSub = CommandSubscription.fromTopic(topic,
+                        authenticatedDevice);
+                if (cmdSub != null) {
+                    final String tenantId = cmdSub.getTenant();
+                    final String deviceId = cmdSub.getDeviceId();
+                    LOG.debug("unsubscribe for [topic:{}] from [tenant:{}, device:{}]",
+                            topic, tenantId, deviceId);
+                    endpoint.unsubscribeAcknowledge(unsubscribe.messageId());
+                    closeCommandConsumer(tenantId, deviceId);
+                    sendTtdEvent(tenantId, deviceId, authenticatedDevice, 0);
+                }
+            } else {
+                LOG.error("unsubscribe for multiple topics then is not allowed");
+            }
+        });
+    }
+
+    private Future<MessageConsumer> createCommandConsumer(final MqttEndpoint mqttEndpoint, final CommandSubscription sub) {
+        return getCommandConnection().getOrCreateCommandConsumer(sub.getTenant(), sub.getDeviceId(), (delivery, message) -> {
+            LOG.trace("command received for [tenant-id: {}, device-id: {}] with delivery ", sub.getTenant(), sub.getDeviceId(),
```

Comment:
this is not really true, it just means that a message has been received. All the code for identifying if it actually is a command and then instantiating a `Command` is in `onCommmandReceived`. IMHO it would indeed be nicer, if this code would be in this handler and `onCommandReceived` would be invoked with a `Command` instead.